google/cloud/spanner/integration_tests/client_integration_test.cc

```diff
@@ -179,58 +179,62 @@ TEST_F(ClientIntegrationTest, TransactionRollback) {
-                        {{"id", Value(3)},
-                         {"fname", Value("test-fname-3")},
-                         {"lname", Value("test-lname-3")}}));
-  EXPECT_STATUS_OK(insert1);
-  auto insert2 = client_->ExecuteSql(
-      txn, SqlStatement("INSERT INTO Singers (SingerId, FirstName, LastName) "
-                        "VALUES (@id, @fname, @lname)",
-                        {{"id", Value(4)},
-                         {"fname", Value("test-fname-4")},
-                         {"lname", Value("test-lname-4")}}));
-  EXPECT_STATUS_OK(insert2);
-
-  auto reader = client_->Read(txn, "Singers", KeySet::All(),
-                              {"SingerId", "FirstName", "LastName"});
-  EXPECT_STATUS_OK(reader);
-
   using RowType = Row<std::int64_t, std::string, std::string>;
   std::vector<RowType> returned_rows;
-  if (reader) {
+
+  Transaction txn = MakeReadOnlyTransaction();
+  for (auto start = std::chrono::steady_clock::now(),
+            deadline = start + std::chrono::minutes(1);
+       start < deadline; start = std::chrono::steady_clock::now()) {
+    txn = MakeReadWriteTransaction();
+    auto insert1 = client_->ExecuteSql(
+        txn, SqlStatement("INSERT INTO Singers (SingerId, FirstName, LastName) "
+                          "VALUES (@id, @fname, @lname)",
+                          {{"id", Value(3)},
+                           {"fname", Value("test-fname-3")},
+                           {"lname", Value("test-lname-3")}}));
+    if (!insert1) continue;
+    auto insert2 = client_->ExecuteSql(
+        txn, SqlStatement("INSERT INTO Singers (SingerId, FirstName, LastName) "
+                          "VALUES (@id, @fname, @lname)",
+                          {{"id", Value(4)},
+                           {"fname", Value("test-fname-4")},
+                           {"lname", Value("test-lname-4")}}));
+    if (!insert2) continue;
+
+    auto reader = client_->Read(txn, "Singers", KeySet::All(),
+                                {"SingerId", "FirstName", "LastName"});
+    if (!reader) continue;
+
+    returned_rows.clear();
     int row_number = 0;
     for (auto& row : reader->Rows<std::int64_t, std::string, std::string>()) {
-      EXPECT_STATUS_OK(row);
       if (!row) break;
       SCOPED_TRACE("Parsing row[" + std::to_string(row_number++) + "]");
       returned_rows.push_back(*std::move(row));
     }
-  }
-  EXPECT_THAT(returned_rows,
-              UnorderedElementsAre(RowType(1, "test-fname-1", "test-lname-1"),
+    ASSERT_THAT(returned_rows, UnorderedElementsAre(
+                                   RowType(1, "test-fname-1", "test-lname-1"),
                                    RowType(2, "test-fname-2", "test-lname-2"),
                                    RowType(3, "test-fname-3", "test-lname-3"),
                                    RowType(4, "test-fname-4", "test-lname-4")));
+    break;
+  }
 
   auto insert_rollback_result = client_->Rollback(txn);
```

Comment:
I'd probably move this inside the loop (before the final `break`) to make `txn` loop local (and so avoid its strange outside-loop initialization).